### PR TITLE
Swift-3-Conversion

### PR DESCRIPTION
**Pull request by supperspidey**
#### Ticket
N/A

#### Ticket Summary
Update `VimeoUpload` code base to Swift 3.

#### Implementation Summary
N/A

#### How to Test
In order for the framework to build, update your `pod` via the command
```
pod install
```

### CODE REVIEW

**Comment by jasonhawkins** [comment]
@supperspidey Go ahead and try pointing to this guy and see what breaks: https://github.com/vimeo/VimeoNetworking/pull/121

**Comment by supperspidey** [comment]
@jasonhawkins it finally works!

File: Examples/VimeoUpload-iOS-OldUpload/VimeoUpload-iOS-OldUpload/ViewControllers/VideoSettingsViewController.swift [hunk 1]
```diff
@@ -71,7 +71,7 @@ class VideoSettingsViewController: UIViewController, UITextFieldDelegate
         
         assert(self.input != nil, "self.input cannot be nil")
         
-        self.edgesForExtendedLayout = .None
+        self.edgesForExtendedLayout = UIRectEdge()
```
Comment (mikew-personal) [inline Examples/VimeoUpload-iOS-OldUpload/VimeoUpload-iOS-OldUpload/ViewControllers/VideoSettingsViewController.swift]: is UIRectEdge() equivalent to .None?

File: Examples/VimeoUpload-iOS/VimeoUpload-iOS/Cells/VideoCell.swift [hunk 2]
```diff
@@ -120,71 +120,71 @@ class VideoCell: UITableViewCell
...
         let constant = CGFloat(1 - progress) * width
         self.progressConstraint.constant = constant
     }
 
-    private func updateState(state: DescriptorState)
+    private func update(state: DescriptorState)
     {
         switch state
         {
         case .Ready:
```
Comment (mikew-personal) [inline Examples/VimeoUpload-iOS/VimeoUpload-iOS/Cells/VideoCell.swift]: shouldn't these enums be changed to lower case?

File: Examples/VimeoUpload-iOS/VimeoUpload-iOS/Cells/VideoCell.swift [hunk 3]
```diff
@@ -120,71 +120,71 @@ class VideoCell: UITableViewCell
...
-            self.updateProgress(0)
-            self.progressView.hidden = false
-            self.deleteButton.setTitle("Cancel", forState: .Normal)
+            self.update(progress: 0)
+            self.progressView.isHidden = false
+            self.deleteButton.setTitle("Cancel", for: UIControlState())
```
Comment (mikew-personal) [inline Examples/VimeoUpload-iOS/VimeoUpload-iOS/Cells/VideoCell.swift]: is UIControlState() equivalent to .Normal?

File: Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift [hunk 4]
```diff
@@ -91,7 +91,7 @@ class VideoSettingsViewController: UIViewController, UITextFieldDelegate
         
         assert(self.input != nil, "self.input cannot be nil")
         
-        self.edgesForExtendedLayout = .None
+        self.edgesForExtendedLayout = UIRectEdge()
```
Comment (mikew-personal) [inline Examples/VimeoUpload-iOS/VimeoUpload-iOS/ViewControllers/VideoSettingsViewController.swift]: is UIRectEdge() equivalent to .None?

File: Podfile [hunk 5]
```diff
@@ -6,14 +6,14 @@ xcodeproj 'Examples/VimeoUpload-iOS-OldUpload/VimeoUpload-iOS-OldUpload.xcodepro
 target 'VimeoUpload' do
     platform :ios, '8.0'
     use_frameworks!
-    pod 'VimeoNetworking', :git => 'git@github.com:vimeo/VimeoNetworking.git', :commit => 'be393b925c3fb523c54b0c59a659e5b3c959858d'
+    pod 'VimeoNetworking', :git => 'git@github.com:vimeo/VimeoNetworking.git', :branch => 'swift-3-part-1'
```
Comment (mikew-personal) [inline Podfile]: I believe the latest branch for VimeoNetworking is swift-3-update

File: VimeoUpload/Descriptor System/Descriptor.swift [hunk 6]
```diff
@@ -35,7 +35,7 @@ public enum DescriptorState: String
     case Finished = "Finished"
 }
 
-public class Descriptor: NSObject, NSCoding
+open class Descriptor: NSObject, NSCoding
```
Comment (mikew-personal) [inline VimeoUpload/Descriptor System/Descriptor.swift]: should all of these open methods be changed back to public?  Do we ever subclass this Descriptor class?

File: VimeoUpload/Descriptor System/DescriptorManagerDelegate.swift [hunk 7]
```diff
@@ -30,21 +30,21 @@ import Foundation
 
 @objc public protocol DescriptorManagerDelegate: class
 {
-    optional func didLoadDescriptors(descriptors descriptors: Set<Descriptor>)
-    optional func didSaveDescriptors(count count: Int)
-    optional func didFailToLoadDescriptor(error error: NSError)
+    @objc optional func didLoadDescriptors(descriptors: Set<Descriptor>)
```
Comment (mikew-personal) [inline VimeoUpload/Descriptor System/DescriptorManagerDelegate.swift]: do we need to add @objc to all of these?

File: VimeoUpload/Descriptor System/KeyedArchiver.swift [hunk 8]
```diff
@@ -55,13 +55,13 @@ public class KeyedArchiver: ArchiverProtocol
     
     // MARK: Utilities
     
-    func archivePath(key key: String) -> String
+    func archivePath(key: String) -> String
     {
-        var URL = NSURL(string: self.basePath)!
+        var URL = Foundation.URL(string: self.basePath)!
```
Comment (mikew-personal) [inline VimeoUpload/Descriptor System/KeyedArchiver.swift]: was this ambiguous?  Do we need to scope URL with Foundation here?

File: VimeoUpload/Descriptor System/ProgressDescriptor.swift [hunk 9]
```diff
@@ -26,7 +26,7 @@
 
 import Foundation
 
-public class ProgressDescriptor: Descriptor
+open class ProgressDescriptor: Descriptor
```
Comment (mikew-personal) [inline VimeoUpload/Descriptor System/ProgressDescriptor.swift]: can we change all these opens to public?  Do we subclass ProgressDescriptor anywhere?

File: VimeoUpload/UIKit/Extensions/NSString+Conversions.swift [hunk 10]
```diff
@@ -61,21 +61,21 @@ public extension NSString
...
     
-    static func stringFromDurationInSeconds(duration: Float64) -> NSString
+    static func stringFromDuration(inSeconds duration: Float64) -> NSString
     {
         let duration = max(duration, 0)
         
         let hours = Int(floor(duration / (60 * 60)))
-        let minutes = hours > 0 ? Int(floor((duration % (60 * 60)) / 60)) : Int(floor(duration / 60))
-        let seconds = Int(floor(duration % 60))
+        let minutes = hours > 0 ? Int(floor((duration.truncatingRemainder(dividingBy: (60 * 60))) / 60)) : Int(floor(duration / 60))
```
Comment (mikew-personal) [inline VimeoUpload/UIKit/Extensions/NSString+Conversions.swift]: are we positive that truncatingRemainder is functionally equivalent to using the modulus operator?

File: VimeoUpload/Upload/Descriptor System/Cameo/CAMUploadDescriptor.swift [hunk 11]
```diff
@@ -12,15 +12,15 @@ import VimeoNetworking
...
-    let videoUrl: NSURL
+    let videoUrl: URL
     let videoSettings: VideoSettings?
-    let thumbnailUrl: NSURL?
+    let thumbnailUrl: URL?
     
     public private(set) var uploadTicket: VIMUploadTicket?
-    private(set) var pictureTicket: VIMThumbnailUploadTicket?
-    private(set) var picture: VIMPicture?
+    private var pictureTicket: VIMThumbnailUploadTicket?
```
Comment (mikew-personal) [inline VimeoUpload/Upload/Descriptor System/Cameo/CAMUploadDescriptor.swift]: These were previously private set and internal get, why are we making the gets private now?

File: VimeoUpload/Upload/Descriptor System/Cameo/VimeoRequestSerializer+Thumbnail.swift [hunk 12]
```diff
@@ -12,56 +12,56 @@ import VimeoNetworking
...
         var error: NSError?
-        let request = self.requestWithMethod("PUT", URLString: destination, parameters: nil, error: &error)
+        let request = self.request(withMethod: "PUT", urlString: destination, parameters: nil, error: &error)
         if let error = error {
-            throw error.errorByAddingDomain(UploadErrorDomain.UploadThumbnail.rawValue)
+            throw error.error(byAddingDomain: UploadErrorDomain.UploadThumbnail.rawValue)
         }
         
-        let asset = AVURLAsset(URL: source)
+        let asset = AVURLAsset(url: source as URL)
```
Comment (mikew-personal) [inline VimeoUpload/Upload/Descriptor System/Cameo/VimeoRequestSerializer+Thumbnail.swift]: why are we casting source to URL if it's already a URL?

File: VimeoUpload/Upload/Model/VideoSettings.swift [hunk 13]
```diff
@@ -26,27 +26,27 @@
 
 import Foundation
 
-public class VideoSettings: NSObject
+open class VideoSettings: NSObject
```
Comment (mikew-personal) [inline VimeoUpload/Upload/Model/VideoSettings.swift]: do these all need to be open?

File: VimeoUpload/VimeoUploader.swift [hunk 14]
```diff
@@ -27,7 +27,7 @@
 import Foundation
 import VimeoNetworking
 
-public class VimeoUploader<T: VideoDescriptor>
+open class VimeoUploader<T: VideoDescriptor>
```
Comment (mikew-personal) [inline VimeoUpload/VimeoUploader.swift]: this should be public.

File: VimeoUpload/Descriptor System/Descriptor.swift (hunk 6 above)
Comment (supperspidey) [inline VimeoUpload/Descriptor System/Descriptor.swift]: I believe `ProgressDescriptor` subclasses `Descriptor`. I can make the class `Descriptor` open but keep its methods and properties `public` if you want.

File: VimeoUpload/Descriptor System/DescriptorManagerDelegate.swift (hunk 7 above)
Comment (supperspidey) [inline VimeoUpload/Descriptor System/DescriptorManagerDelegate.swift]: I believe we need to. Optional protocol methods only exist in Objective-C but not in Swift.

File: VimeoUpload/Descriptor System/ProgressDescriptor.swift (hunk 9 above)
Comment (supperspidey) [inline VimeoUpload/Descriptor System/ProgressDescriptor.swift]: `ProgressDescriptor` is being subclassed by `UploadDescriptor` and `OldUploadDescriptor`, so we should keep it `open` for now.

File: VimeoUpload/VimeoUploader.swift (hunk 14 above)
Comment (supperspidey) [inline VimeoUpload/VimeoUploader.swift]: Right now there are two classes in the example projects that subclass `VimeoUploader`: `NewVimeoUploader` and `OldVimeoUploader`. Should we change the scope of `VimeoUploader` to `public` then?

File: VimeoUpload/UIKit/Extensions/NSString+Conversions.swift (hunk 10 above)
Comment (supperspidey) [inline VimeoUpload/UIKit/Extensions/NSString+Conversions.swift]: I wondered the same thing too, so I googled the difference between using `%` and `truncatingRemainder`. Turns out in Swift 2, we could compute the remainder of `float`s via `%` (Is float remainder computation a real thing?!). Starting with Swift 3, that operation has to be performed via a call to `truncatingRemainder`.

http://stackoverflow.com/questions/40495301/what-does-is-unavailable-use-truncatingremainder-instead-mean

**Comment by supperspidey** [comment]
build this please

**Comment by cameo-github** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/38133a9967b6182c5728855be13b7f5ae51fc518/VimeoUpload/Upload/Extensions/AVAsset+Extensions.swift#L32">VimeoUpload/Upload/Extensions/AVAsset+Extensions.swift#L32</a></strong>: extraneous duplicate parameter name; ‘completion’ already has an argument label  <br /><code>
    func approximateFileSize(completion completion: @escaping FloatBlock)
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.310 (0.312) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.362 (0.365) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoUpload/158)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Pull request merged**